File: Leetcode/Arrays/find-minimum-in-rotated-sorted-array-153.js
//153. Find Minimum in Rotated Sorted Array

function findMin(array) {
  let lower = 0;
  let upper = array.length - 1;

  while (lower + 1 < upper) {
    let mid = Math.floor(lower + (upper - lower) / 2);
    let midVal = array[mid];
    let rightVal = array[mid + 1];
    let farthestRightVal = array[upper];
    
    if (midVal > rightVal) {
      return rightVal;
    } else if ( rightVal > midVal && farthestRightVal > midVal) {
      upper = mid;
    } else {
      lower = mid;
    }
  }

  return array[lower] < array[upper] ? array[lower] : array[upper];
}

// Mental Model - Find the Ditch. We know the values are ascending and 
// that there is only one point, "the ditch", where the next value will
// descend

File: Leetcode/Binary Search/find-minimum-in-rotated-sorted-array-153.rb
# Leetcode: #153 Find Minimum in Rotated Sorted Array

def find_min(nums, left=0, right=nums.length - 1)
  return nums[left] if left >= left

  mid = left + (right - left) / 2
  mid_val = nums[mid]

  if mid_val > nums[right]
    return find_min(nums, mid + 1, right)
  else
    # we do not do mid - 1 here because
    # mid COULD be the pivot. We only know
    # that elements to the right of it
    # CANNOT be the pivot
    return find_min(nums, left, mid)
  end
end
